### PR TITLE
MINOR: improve logging

### PR DESCRIPTION
**Pull request by mjsax**
Simply reading log by sorting partitions.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 1]
```diff
@@ -215,7 +215,7 @@ public String protocolType() {
 
     @Override
     protected JoinGroupRequestData.JoinGroupRequestProtocolCollection metadata() {
-        log.debug("Joining group with current subscription: {}", subscriptions.subscription());
+        log.debug("Joining group with current subscription: {}", Utils.join(subscriptions.subscription().stream().sorted().toArray(), ", "));
```
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Could we just keep the subscription sorted to begin with? (ie store in a sorted data structure)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 2]
```diff
@@ -288,7 +288,7 @@ private Exception invokeOnAssignment(final ConsumerPartitionAssignor assignor, f
     }
 
     private Exception invokePartitionsAssigned(final Set<TopicPartition> assignedPartitions) {
-        log.info("Adding newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
+        log.info("Adding newly assigned partitions: {}", Utils.join(assignedPartitions.stream().sorted().toArray(), ", "));
```
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: FYI `TopicPartition` doesn't implement `Comparable` so you can't use `.stream().sorted()`

**Review by ableegoldman — COMMENTED** [review]
Need to fix the sorting mechanism but otherwise LGTM

Or maybe we should just have `TopicPartition` implement `Comparable<>`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 3]
```diff
@@ -397,10 +397,10 @@ protected void onJoinComplete(int generation,
...
                     "\tAdded partitions (assigned - owned):       {}\n" +
                     "\tRevoked partitions (owned - assigned):     {}\n",
-                assignedPartitions,
-                ownedPartitions,
-                addedPartitions,
-                revokedPartitions
+                Utils.join(assignedPartitions.stream().sorted().toArray(), ", "),
+                Utils.join(ownedPartitions.stream().sorted().toArray(), ", "),
+                Utils.join(addedPartitions.stream().sorted().toArray(), ", "),
+                Utils.join(revokedPartitions.stream().sorted().toArray(), ", ")
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Actually, the `assignedPartitions`, `addedPartitions`, and `revokedPartitions`, and the following `invokePartitionsRevoked`, `invokePartitionsAssigned` are coming from the `assignment.partitions()`(or the `assignedPartitions`). We can sort it first, and the rest of the partitions will be sorted.

Just the `ownedPartitions` might still need to sort itself.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 2 above)
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Good point!

**Review by showuon — COMMENTED** [review]
@mjsax , thanks for the improvement, left some comments. Thanks.

**Comment by mjsax** [comment]
Finally updated this PR.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 4]
```diff
@@ -1516,4 +1527,20 @@ RebalanceProtocol getProtocol() {
     boolean poll(Timer timer) {
         return poll(timer, true);
     }
+
+
+
+    final static class TopicPartitionComparator implements Comparator<TopicPartition>, Serializable {
```
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Need to implement `Serializable` to make spotbug happy 🤷

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 5]
```diff
@@ -1516,4 +1527,20 @@ RebalanceProtocol getProtocol() {
...
+        @Override
+        public int compare(TopicPartition topicPartition1, TopicPartition topicPartition2) {
+            String topic1 = topicPartition1.topic();
+            String topic2 = topicPartition2.topic();
+
+            if (topic1.equals(topic2)) {
+                return topicPartition1.partition() - topicPartition2.partition();
+            } else {
+                return topicPartition1.topic().compareTo(topicPartition2.topic());
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: nit: `topic1.compareTo(topic2)`

**Review by showuon — COMMENTED** [review]
Thanks for the update! And the tests for partitionComparator. Overall LGTM! Left 1 minor comment. Thanks.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 5 above)
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Good catch!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 6]
```diff
@@ -1516,4 +1527,20 @@ RebalanceProtocol getProtocol() {
     boolean poll(Timer timer) {
         return poll(timer, true);
     }
+
+
```
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: nit: extra lines

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 4 above)
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: nit: seems a bit weird to have this in the `ConsumerCoordinator`, would make more sense in a utility class or something related to TopicPartitions -- obviously the actual TopicPartition class would be ideal but unfortunately that would require a KIP, not sure what the second best choice would be but maybe `Topic` or `Utils` or even `CollectionUtils` (under the clients/common package)

**Review by ableegoldman — APPROVED** [review]
few nits, but otherwise LGTM

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 4 above)
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Good point @ableegoldman ! Actually, there's also another partition comparator in [AbstractStickyAssignor.java](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java#L1021) . We could make them into util class or `TopicPartitions` class if needed.

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Ack. Just pushed an update with the refactoring, adding a `Utils` class.

**Review by showuon — COMMENTED** [review]
Cool! Looks great!
But looks like checkstyle doesn't like your indent...
```
09:20:06  [ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11584/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java:27:3: 'class def modifier' has incorrect indentation level 2, expected level should be 4. [Indentation]
09:20:06  [ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-11584/clients/src/main/java/org/apache/kafka/clients/consumer/internals/Utils.java:28:7: 'member def modifier' has incorrect indentation level 6, expected level should be 8. [Indentation]
...
```

**Review by showuon — APPROVED** [review]
LGTM! Thanks.
PS. There are many flaky tests recently. I'll fix them when available. (I hate them!)

**Comment by mjsax** [comment]
Thanks -- it's also ok to first just file Jira's for flaky tests -- I often don't know the code base (core/connect)... Of course, trying to fix them right away is even better.

**Pull request merged**